### DISCUSSION TITLE
Question: compiling tools

### DISCUSSION THREAD

**Discussion by VictorEijkhout** [OrderN/CONQUEST-release#159] [Compiling] [answered]
```
mpif90 -O3 -I/usr/local/include -c radial_xc_LibXC_v4_module.f90
ifort: command line warning #10006: ignoring unknown option '-fallow-argument-mismatch'
radial_xc_LibXC_v4_module.f90(4): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [XC_F90_TYPES_M]
  use xc_f90_types_m
------^
```

I'm not finding this module anywhere in the source. How do I proceed?

**Comment by smujahed** [comment]
Hi @VictorEijkhout 

Would you be able to upload your `system.make` file here? There may be an issue with your LibXC, either at the point it is linked OR LibXC may have been compiled with a different compiler. The end of the first line looks like it's not finding the LibXC library, so appears to be a linking issue, do you know the path to your LibXC library?

Sorry I can't be of more help right now, but with the `system.make` file and some more questions I'm sure we can sort this out. 

All the best,
Shereif.

**Comment by VictorEijkhout** [comment]
I'm trying to use the internal XC library as it says on your web page:

```
#

# Set compilers
FC=mpif90
F77=mpif77

# Linking flags
LINKFLAGS=
## -L/usr/local/lib
ARFLAGS=

# Compilation flags
# NB for gcc10 you need to add -fallow-argument-mismatch
COMPFLAGS= -O2 -g $(XC_COMPFLAGS)
COMPFLAGS_F77= $(COMPFLAGS)

# Set BLAS and LAPACK libraries
# MacOS X
# BLAS= -lvecLibFort
# Intel MKL use the Intel tool
BLAS = -mkl
# Generic
# BLAS= -llapack -lblas

# Full library call; remove scalapack if using dummy diag module
LIBS= $(FFT_LIB) $(XC_LIB) $(BLAS) \
    -L${TACC_MKL_LIB} \
    -lmkl_scalapack_lp64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -lmkl_blacs_intelmpi_lp64 \-lpthread
##  VLE -lscalapack

# LibXC compatibility (LibXC below) or Conquest XC library

# Conquest XC library
XC_LIBRARY = CQ
#XC_LIB =
#XC_COMPFLAGS =

# LibXC compatibility
# Choose LibXC version: v4 or v5
# XC_LIBRARY = LibXC_v4
# XC_LIBRARY = LibXC_v5
# XC_LIB = -lxcf90 -lxc
# XC_COMPFLAGS = -I/usr/local/include

# Set FFT library
FFT_LIB= -L${LMOD_FFTW3_LIB} -lfftw3
FFT_OBJ=fft_fftw3.o

# Matrix multiplication kernel type
MULT_KERN = default
# Use dummy DiagModule or not
DIAG_DUMMY =
```

**Comment by smujahed** [comment]
OK, so it looks like `XC_COMPFLAGS` was left commented out. Could you uncomment that, `make clean`, then `make` and report back? It would also be useful to know a little more about the CONQUEST code version (commit hash), compiler and libraries (with versions) used to compile.

Just for clarification, are you compiling CONQUEST's main code (src/) or just the tools (tools/)? My apologies, the issue title could be interpreted either way and we can better help with this clarification.

**Comment by VictorEijkhout** [comment]
I'm compiling main code & tools. the `XC_COMPFLAGS` only includes `/usr/local/include`. I really doubt that Conquest fortran modules are there. In fact, I can't find the `xc_f90_lib_m` module anywhere in the source tree.

Code version: I'm using the repo version from a couple of days ago. The top entry in the `git log` is 3a6e1f697e0bd3fd1101715e56bf3c7b7a7fde63

**Comment by davidbowler** [comment]
You need to set the following in system.make:

`XC_LIBRARY = CQ`
`XC_LIB = `
`XC_COMPFLAGS = `

Then please try `make clean` and then `make`

**Comment by davidbowler** [comment]
LibXC is not something that we provide with Conquest, but it is easily downloaded and compile (and is often on large HPC installations or in distributions).  Once you have identified the LibXC location, the include path (which you need to set for your system) should pick up the module files.

**Comment by davidbowler** [comment]
(I have changed this to a discussion as I don't think that it's a bug or an issue; we will try to make the compiling instructions clearer.)

**Comment by VictorEijkhout** [comment]
I have installed `libcx`. Q: your makefile talks about a `v4` and `v5`, but `libcx` is on version 6.whatever. Do I need to use an earlier version 5 instead?

Also: Im specifying `v5` but the compilation process still has:
```
radial_xc_LibXC_v4_module.f90(4): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [XC_F90_TYPES_M]
  use xc_f90_types_m
------^
```
Should it really compile a `v4` file?

**Comment by davidbowler** [comment]
Conquest only supports LibXC v4 or v5 (v6 is relatively recent).

I don't understand why your make is picking up v4 (though this will depend on the values you have set for the XC variables in `system.make`).  For now, to test that you can compile simply, please try what I suggested above (set `XC_LIBRARY = CQ` and the other variables to blank and make sure that these are the *only* XC variables defined in `system.make`; then try `make clean` followed by `make`)

If you could attach your `system.make` file that might help.

**Reply by VictorEijkhout** [reply]
> I don't understand why your make is picking up v4 

Because `tools/BasisGeneration/deps.obj` is full of references to that?

**Comment by VictorEijkhout** [comment]
I was trying to link XClib:

```
COMPFLAGS= -O2 -g $(XC_COMPFLAGS)
COMPFLAGS_F77= $(COMPFLAGS)

# Set BLAS and LAPACK libraries
# MacOS X
# BLAS= -lvecLibFort
# Intel MKL use the Intel tool
BLAS = -mkl
# Generic
# BLAS= -llapack -lblas

# Full library call; remove scalapack if using dummy diag module
LIBS= $(FFT_LIB) $(XC_LIB) $(BLAS) \
    -L${TACC_MKL_LIB} \
    -lmkl_scalapack_lp64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -lmkl_blacs_intelmpi_lp64 -\lpthread
##  VLE -lscalapack

# LibXC compatibility (LibXC below) or Conquest XC library

# Conquest XC library
#XC_LIBRARY = LibXC_v5
XC_LIBRARY = CQ
XC_LIB = -L${LMOD_LIBXC_LIB} -lxcf90 -lxc
XC_COMPFLAGS = -I${LMOD_LIBXC_INC}
```

where

```
[staff conquest:291] env | grep lmod_libx
LMOD_LIBXC_BIN=/work2/00434/eijkhout/libxc/installation-6.1.0-skx-intel/bin
LMOD_LIBXC_LIB=/work2/00434/eijkhout/libxc/installation-6.1.0-skx-intel/lib64
LMOD_LIBXC_INC=/work2/00434/eijkhout/libxc/installation-6.1.0-skx-intel/include
LMOD_LIBXC_DIR=/work2/00434/eijkhout/libxc/installation-6.1.0-skx-intel
```

Using blank variables:

```
# Conquest XC library
#XC_LIBRARY = LibXC_v5
XC_LIBRARY = CQ
XC_LIB =
## -L${LMOD_LIBXC_LIB} -lxcf90 -lxc
XC_COMPFLAGS =
## -I${LMOD_LIBXC_INC}
```

gives

```
mpif90 -O3 -I/usr/local/include -c radial_xc_LibXC_v4_module.f90
ifort: command line warning #10006: ignoring unknown option '-fallow-argument-mismatch'
radial_xc_LibXC_v4_module.f90(4): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [XC_F90_TYPES_M]
  use xc_f90_types_m
------^
```

and there are *no* references to v5.

**Reply by smujahed** [reply]
In both the first code block and the blank code block your flag is set to `XC_LIBRARY = CQ`, which will cause compilation to use the internal CONQUEST XC library. If this is because you plan to compile using the internal CONQUEST XC library, the last code block with blank `XC_LIB` and `XC_COMPFLAGS` arguments is correct. Taking that with your previous comment about `tools/BasisGeneration/deps.obj` being full of references to LibXC_v4, my suggestion would be to try `make very_clean` and then try `make` again. You may need to `make clean` in the tools directory too. Please report back on the results of that compilation. 

In the first code block there are a few issues. First, `XC_LIBRARY = CQ` will force compilation with CQ internal XC. The `XC_LIB = -L${LMOD_LIBXC_LIB} -lxcf90 -lxc` contains the flag `-lxc` which is a flag to link LibXC (and possibly `-lxcf90`). In fact, the entire line appears to link to LibXC. The issue is the same on the following line `XC_COMPFLAGS`. So the issue in that it appears to be confused whether to use CQ or LibXC. If the intention is to link LibXC, then the line `#XC_LIBRARY = LibXC_v5` should be uncommented and the other reference to `XC_LIBRARY = CQ` commented out. Then, the second issue with that compile is that you seem to be linking to LibXC v6, which is not currently supported (as per @davidbowler comment above). In the instance you want to compile with LibXC, please download and install either a v5 or v4 version from the [TDDFT repo](https://tddft.org/programs/libxc/download/previous/). I personally have no recommendation as to which minor version to use, Dave may though, but generally I'd go for the latest v5.2.3 or v4.3.4. As always, `make clean` or `make very_clean` and recompile.

I hope this helps resolve your compilation woes!

All the best,
Shereif.

**Reply by VictorEijkhout** [reply]
1. `very_clean` makes no difference. The `tools/BasisGeneration/deps.obj` keeps having `v4` references. In fact, they are there even if I make a new git clone.

2. Issues, yes, well, .... I've tried about every combination. If I don't use `CQ` but `libv5` then I see v5 appearing in the link lines, and v4 in the compilation errors.

So: should there be v4 references in that tool/deps file?

**Comment by VictorEijkhout** [comment, marked as answer]
And the missing piece of the puzzle:

You Have To Also Put `system.make` In The Tools/Whatever Directory

Updated compile instructions would be very much appreciated.

**Comment by davidbowler** [comment]
Hi Victor.  I'm glad that you found the problem.  In the Conquest manual, under Installation, the sentence "The ion file generation code is compiled using the same options required for the main code." could be made a little clearer - we'll update that in a future release.

If you have future issues, please do tell us which directory your error is and include all available information (system.make file, full output from compilation) as an attachment as is standard practice for most community supported codes.

I hope that you find Conquest useful.

**Comment by davidbowler** [comment]
I have raised an issue to simplify the `system.make` files (the same file is needed in three places so should be copied or imported).  I have also tested LibXC v6.1 which is compatible with the Conquest LibXC5 module, and have raised another issue to note this in the naming or compilation instructions.
